File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums, int i, vector<int>& dp)
    {
        if(i<0)
            return 0;
        if(dp[i] != -1)
            return dp[i];
        return dp[i] = max(nums[i] + rob(nums, i-2, dp), rob(nums, i-1, dp));
    }
    int rob(vector<int>& nums) {
        
        vector<int>dp(nums.size(), -1);
        return rob(nums, nums.size()-1, dp);
    }
};

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        n = len(nums)
        dp = {}
        def rob(index, prev):
            
            if index == n:
                return 0
            
            if (index, prev) in dp:
                return dp[index, prev]
            
            if prev != -1 and prev == index - 1: 
                dp[index, prev] = rob(index + 1, prev)
            else:
                dp[index, prev] = max(rob(index + 1, prev),  rob(index + 1, index) + nums[index])
            
            return dp[index, prev]
        
        return rob(0, -1)
    
        
        
        